src/ripple/app/misc/impl/AMM.cpp

```diff
@@ -0,0 +1,248 @@
+    Currency const& cur1,
+    Currency const& cur2,
+    AccountID const& ammAccount,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    auto const lptIss = ammLPTIssue(cur1, cur2, ammAccount);
+    return accountHolds(
+        view,
+        lpAccount,
+        lptIss.currency,
+        lptIss.account,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        j);
+}
+
+STAmount
+ammLPHolds(
+    ReadView const& view,
+    SLE const& ammSle,
+    AccountID const& lpAccount,
+    beast::Journal const j)
+{
+    return ammLPHolds(
+        view,
+        ammSle[sfAsset].currency,
+        ammSle[sfAsset2].currency,
+        ammSle[sfAMMAccount],
+        lpAccount,
+        j);
+}
+
+bool
+isFrozen(ReadView const& view, STAmount const& a)
+{
+    return !a.native() && isGlobalFrozen(view, a.getIssuer());
+}
+
+TER
+requireAuth(ReadView const& view, Issue const& issue, AccountID const& account)
+{
+    if (isXRP(issue) || issue.account == account)
+        return tesSUCCESS;
+    if (auto const issuerAccount = view.read(keylet::account(issue.account));
+        issuerAccount && (*issuerAccount)[sfFlags] & lsfRequireAuth)
+    {
+        if (auto const trustLine =
+                view.read(keylet::line(account, issue.account, issue.currency));
+            trustLine)
+            return !((*trustLine)[sfFlags] &
+                     ((account > issue.account) ? lsfLowAuth : lsfHighAuth))
+                ? TER{tecNO_AUTH}
+                : tesSUCCESS;
+        return TER{tecNO_LINE};
+    }
+
+    return tesSUCCESS;
+}
+
+std::uint16_t
+getTradingFee(ReadView const& view, SLE const& ammSle, AccountID const& account)
+{
+    using namespace std::chrono;
+    if (ammSle.isFieldPresent(sfAuctionSlot))
+    {
+        auto const& auctionSlot =
+            static_cast<STObject const&>(ammSle.peekAtField(sfAuctionSlot));
+        if (auto const expiration = auctionSlot[~sfExpiration])
+        {
+            auto const notExpired =
+                duration_cast<seconds>(
+                    view.info().parentCloseTime.time_since_epoch())
+                    .count() < expiration;
+            if (auctionSlot[~sfAccount] == account && notExpired)
+                return auctionSlot[sfDiscountedFee];
+            if (auctionSlot.isFieldPresent(sfAuthAccounts))
+            {
+                for (auto const& acct :
+                     auctionSlot.getFieldArray(sfAuthAccounts))
+                    if (acct[~sfAccount] == account && notExpired)
+                        return auctionSlot[sfDiscountedFee];
+            }
+        }
+    }
+    return ammSle[sfTradingFee];
+}
+
+TER
+ammSend(
+    ApplyView& view,
+    AccountID const& from,
+    AccountID const& to,
+    STAmount const& amount,
+    beast::Journal j)
+{
+    if (isXRP(amount))
+        return accountSend(view, from, to, amount, j);
+
+    auto const issuer = amount.getIssuer();
+
+    if (from == issuer || to == issuer || issuer == noAccount())
```

Comment:
> The code is taken from `accountSend()` and is essentially the same except that the IOU transfer fee is ignored. I could have added a `waveFee` flag to `accountSend` but I wanted to avoid changing a fundamental API.

IMHO, changing an API is far preferable to code duplication in a function as important as this one.

Usually, we always put any `beast::Journal` parameter as the last parameter. The exception is when there's a default value. I think it would be perfectly fine to define
```
[[nodiscard]] TER
accountSend(
    ApplyView& view,
    AccountID const& from,
    AccountID const& to,
    const STAmount& saAmount,
    beast::Journal j,
    bool waiveFee = false);
```
and go from there.